core/dbt/adapters/base/impl.py

```diff
@@ -1104,3 +1134,30 @@ def post_model_hook(self, config: Mapping[str, Any], context: Any) -> None:
         The second parameter is the value returned by pre_mdoel_hook.
         """
         pass
+
+
+def catch_as_completed(
+    futures  # typing: List[Future[agate.Table]]
+) -> Tuple[agate.Table, List[Exception]]:
+
+    catalogs: agate.Table = agate.Table(rows=[])
+    exceptions: List[Exception] = []
+
+    for future in as_completed(futures):
+        exc = future.exception()
+        # we want to re-raise on ctrl+c and BaseException
+        if exc is None:
+            catalog = future.result()
+            catalogs = agate.Table.merge([catalogs, catalog])
```

Comment:
Agate seems unhappy about this on Snowflake. I see:
```
2020-02-04 02:07:41,316840 (MainThread): Tables contain columns with the same names, but different types.
2020-02-04 02:07:41,322568 (MainThread): Traceback (most recent call last):
  File "/Users/drew/fishtown/dbt/core/dbt/main.py", line 81, in main
    results, succeeded = handle_and_check(args)
  File "/Users/drew/fishtown/dbt/core/dbt/main.py", line 159, in handle_and_check
    task, res = run_from_args(parsed)
  File "/Users/drew/fishtown/dbt/core/dbt/main.py", line 212, in run_from_args
    results = task.run()
  File "/Users/drew/fishtown/dbt/core/dbt/task/generate.py", line 214, in run
    catalog_table, exceptions = adapter.get_catalog(self.manifest)
  File "/Users/drew/fishtown/dbt/core/dbt/adapters/base/impl.py", line 1061, in get_catalog
    catalogs, exceptions = catch_as_completed(futures)
  File "/Users/drew/fishtown/dbt/core/dbt/adapters/base/impl.py", line 1151, in catch_as_completed
    catalogs = agate.Table.merge([catalogs, catalog])
  File "/Users/drew/fishtown/dbt/env/lib/python3.7/site-packages/agate/table/merge.py", line 45, in merge
    raise DataTypeError('Tables contain columns with the same names, but different types.')
agate.exceptions.DataTypeError: Tables contain columns with the same names, but different types.
```

In this case, the error is super annoying. One `information_schema` contains a table with a clustering key defined (of type `string`). The other `information_schema` that gets merged in does not have _any_ clustering keys on any tables, so every value in the `stats:clustering_key:value` columns is `None`, which I guess Agate assigned to an integery type?

I ran some debug code to see what was going wrong here, and I saw that the types for `catalog` and `catalogs` were:
```
stats:clustering_key:value': {<agate.data_types.number.Number object at 0x109d26a10>,
                                <agate.data_types.text.Text object at 0x10a02c390>}
```

While I saw this on Snowflake, I have to imagine we'd see similar things on other databases. Any good ideas about what to do about this one? My (not ideal) thinking is that we could push the raw results into a list, then convert that raw list of lists back into a Table...

Reply:
I have a fix for this (roll my own merge function that tracks null-ness). I think it's actually pretty ok.